src/lib/y2storage/used_storage_features.rb

```diff
@@ -0,0 +1,208 @@
+
+        # Other filesystems
+        UF_EXT2:          "e2fsprogs",
+        UF_EXT3:          "e2fsprogs",
+        UF_EXT4:          "e2fsprogs",
+        UF_XFS:           "xfsprogs",
+        UF_REISERFS:      "reiserfs",
+        UF_NFS:           "nfs-client",
+        UF_NFS4:          "nfs-client",
+        UF_NTFS:          ["ntfs-3g", "ntfsprogs"],
+        UF_VFAT:          "dosfstools",
+
+        # Crypto technologies
+        UF_LUKS:          "cryptsetup",
+        UF_CRYPT_TWOFISH: "cryptsetup",
+
+        # Data transport methods
+        UF_ISCSI:         "open-iscsi",
+        UF_FCOE:          "fcoe-utils",
+        UF_FC:            nil,
+
+        # Other
+        UF_QUOTA:         "quota",
+        UF_BCACHE:        "bcache-tools",
+
+        # FIXME: This is not related to the devicegraph, so libstorage doesn't
+        # return it yet. The "efibootmgr" package is available in the inst-sys
+        # anyway, and it is only needed in very exotic cases in the installed
+        # system: Only if the user wishes to delete a partition (e.g. his
+        # Windows partition) the EFI bootloader could boot from. Adding a
+        # partition to it is handled by yast-bootloader in the inst-sys.
+        UF_EFIBOOT:       "efibootmgr"
+      }
+    # configurable part ends here
+    #======================================================================
+
+    #
+    #----------------------------------------------------------------------
+    #
+
+    def initialize(devicegraph)
+      @devicegraph = devicegraph
+    end
+
+    # Collect storage features and return a feature list
+    # (a list containing :UF_xy symbols). The list may be empty.
+    #
+    # @return [Array<Symbol>] feature list
+    #
+    def collect_features
+      return [] if @devicegraph.nil?
+      feature_bits = @devicegraph.used_features
+      features_dumped = false
+      features = []
+
+      FEATURE_PACKAGES.each_key do |feature|
+        begin
+          mask = bitmask(feature)
+          if (feature_bits & mask) == mask
+            features << feature
+            log.info("Detected feature #{feature}")
+          end
+        rescue NameError => err
+          if err.name == feature
+            log.warn("WARNING: Packages configured for unknown feature :#{feature}")
+            log.info("Features known to libstorage: #{libstorage_features.sort}") unless features_dumped
+            features_dumped = true
+          else
+            log.error("Error: #{err}")
+            raise
+          end
+        end
+      end
+
+      log.info("Storage features used: #{features}")
+      features
+    end
+
+    # Return a list of software packages required for the storage features.
+    # Uses 'features' if non-nil, otherwise collects the features with
+    # collect_features.
+    #
+    # @param [Array<Symbol>] feature list or nil
+    # @return [Array<Symbol>] package list
+    #
+    def feature_packages(features = nil)
```

Comment:
So if a list of features is passed, this acts as some kind of class method (the internal state of the object is irrelevant). But if the parameter is nil, then the devicegraph pointed by the object becomes relevant. This looks like two different methods to me.

```ruby
# Class method. As usual, I suck at naming methods
def self.packages_for(features)
  # Pretty much what we have
end
# instance method
def feature_packages
  self.class.packages_for(collect_features)
end
```